lua/cfc_time/server/time/sv_time_tracking.lua

```diff
@@ -0,0 +1,81 @@
+CFCTime.ctime = {}
+local ctime = CFCTime.ctime
+
+local function now()
+    return os.time()
+end
+
+-- <steamID64> = { joined = <timestamp>, departed = <timestamp> | nil, initialTime = <float> }
+ctime.pendingUpdates = {}
+ctime.updateTimerName = "CFC_Time_UpdateTimer"
+ctime.lastUpdate = now()
+
+function ctime:updateTimes()
+    local batch = {}
+    local now = now()
+
+    for steamId, data in pairs( self.pendingUpdates ) do
+        local isValid = true
+        local joined = data.joined
+        local departed = data.departed
+        local initialTime = data.initialTime
+
+        if departed < self.lastUpdate then
+            self.pendingUpdates[steamId] = nil
+            isValid = false
+        end
+
+        if isValid then
+            local sessionTime = now - joined
+            local newTime = initialTime + sessionTime
+
+            batch[steamId] = newTime
+        end
+    end
+
+    CFCTime.SQL:UpdateTimeBatch( batch )
+    ctime.lastUpdate = now
+end
+
+function ctime:startTimer()
+    timer.Create(
+        self.updateTimerName,
+        self.Config.updateInterval,
+        0,
+        function() ctime:updateTimes() end
+    )
+end
+
+function ctime:stopTimer()
+    timer.Remove( self.updateTimerName )
+end
+
+function ctime:initPlayer( ply )
+    local now = now()
+    local steamId = ply:SteamID64()
+    local initialTime = CFCTime.SQL:getTime( steamId )
+
+    self.pendingUpdates[steamId] = {
+        joined = now,
+        initialTime = initialTime
+    }
+end
+
+function ctime:cleanupPlayer( ply )
+    -- TODO: Verify bug report from the wiki: https://wiki.facepunch.com/gmod/GM:PlayerDisconnected
+    local steamId = ply:SteamID64()
+    self.pendingUpdates[steamId].departed = now()
+end
+
+hook.Add( "Think", "CFC_Time_Init", function()
+    ctime:startTimer()
+    hook.Remove( "Think", "CFC_Time_Init" )
```

Comment:
Imo, its generally good practice to remove the hook before doing anything, in case something goes wrong.
Might just be a preference thing tho

Reply:
yeah, I agree